### DISCUSSION TITLE
Loading Python Exported Model into TorchSharp

### DISCUSSION THREAD

**Discussion by jimquittenton** [xamarin/TorchSharp#585] [General]
Hi, 
I'm new to TorchSharp and am having trouble loading a python trained ResNet18 model. I've been following this article: https://github.com/dotnet/TorchSharp/blob/main/docfx/articles/saveload.md and have exported my python model using the 'save_state_dict' function in this script: https://github.com/dotnet/TorchSharp/blob/main/src/Python/exportsd.py . 

In TorchSharp I have copied the ResNet model from https://github.com/dotnet/TorchSharpExamples/blob/main/src/CSharp/Models/ResNet.cs and then call the following:

```
int numClasses = 3;
ResNet myModel = ResNet.ResNet18(numClasses);
myModel.to(DeviceType.CPU);
myModel.load(mPath);
```
The load() line throws an exception with message `Mismatched module state names: the target modules does not have a submodule or buffer named 'conv1.weight'`.

If I examine the state_dict from 'myModel' prior to load(), it contains entries like:
```
{[layers.conv2d-first.weight, {TorchSharp.Modules.Parameter}]}
{[layers.bnrm2d-first.weight, {TorchSharp.Modules.Parameter}]}
{[layers.bnrm2d-first.bias, {TorchSharp.Modules.Parameter}]}
{[layers.bnrm2d-first.running_mean, {TorchSharp.torch.Tensor}]}
{[layers.bnrm2d-first.running_var, {TorchSharp.torch.Tensor}]}
{[layers.bnrm2d-first.num_batches_tracked, {TorchSharp.torch.Tensor}]}
{[layers.blck-64-0.layers.blck-64-0-conv2d-1.weight, {TorchSharp.Modules.Parameter}]}
{[layers.blck-64-0.layers.blck-64-0-bnrm2d-1.weight, {TorchSharp.Modules.Parameter}]}
{[layers.blck-64-0.layers.blck-64-0-bnrm2d-1.bias, {TorchSharp.Modules.Parameter}]}
```
whereas the corresponding entries prior to saving from python are:
```
conv1.weight     torch.Size([64, 3, 7, 7])
bn1.weight       torch.Size([64])
bn1.bias         torch.Size([64])
bn1.running_mean         torch.Size([64])
bn1.running_var          torch.Size([64])
bn1.num_batches_tracked          torch.Size([])
layer1.0.conv1.weight    torch.Size([64, 64, 3, 3])
layer1.0.bn1.weight      torch.Size([64])
layer1.0.bn1.bias        torch.Size([64])
```

I tried amending the `ResNet.cs` code to reflect the python names, but could not get them to exactly match. 

I also tried calling load() with strict=false `myModel.load(mPath, false);`. This seemed to get past the Mismatched names exception, but throws another exception with message `Too many bytes in what should have been a 7 bit encoded Int32.`

I've been struggling with this for a couple of days now so would really appreciate any help you guys could offer.

Thanks
Jim

**Comment by NiklasGustafsson** [comment]
Hi @jimquittenton,

Passing 'strict=false' is not a solution here, even if it didn't throw an exception (that seems like a bug to me), it would just end up with a partially loaded model, or nothing loaded at all.

A particular network architecture can be constructed many ways, and the names given to the various layers will depend on the details of how it is constructed. You will have __exactly__ match the Python construction logic in the TorchSharp code, or the weights won't have the same names, as you have discovered.

The examples code is tricky, because it is used to construct a number of different ResNet architectures and uses loops and such to construct it. Another trickiness of the example is that it is built to serve as an example and uses the "toy" CIFAR data set. The input image size and the number of classes may not match what your ResNet18 model was trained on (if I'm stating the obvious, please excuse me).

The exception could be because the C# tensors may have different dimensions that the serialized ones, I would have to take a look at that. We may need to change the serialization format to catch issues like that, if that is indeed the case.

**Comment by NiklasGustafsson** [comment]
BTW, I released 0.96.6 on Saturday morning. It has the correct ResNetNN (and a few others) architectures.

**Comment by jimquittenton** [comment]
Thank you Niklas, I'll try this out.
